[SILGenCleanup] Complete lifetimes of defs backwards reachable from dead-end blocks.

nate-chandler opened the pull request:
For a function to have complete lifetimes, the lifetime of every def in the function which is backwards-reachable from a dead-end block must be completed.

Previously, every def in every block which appears in the post-order of the function's blocks is completed.  This was not enough: defs in "unreachable blocks" (i.e. those which aren't forwards-reachable from the function entry) must be completed too, and such blocks do not appear in the function's post-order.

Here, defs in unreachable blocks are be completed too.  Such defs must also be completed in relative post-order.  To do this, roots for the non-entry post-orders must be found.

rdar://141197164

nate-chandler committed:
[NFC] SIL: Find unreachable-terminated blocks.

Add a helper similar to findExitingBlocks.

nate-chandler committed:
[NFC] SILGenCleanup: Added logging.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
[Gardening] Detypo'd comment.

nate-chandler committed:
[NFC] ReachableBlocks: Extract compute method.

In preparation for adding more users.

nate-chandler committed:
[NFC] CLF: Compute reachable blocks only.

Don't form a set of unreachable blocks, we only need to check whether
any given block is unreachable, which can be done via
`ReachableBlocks::isVisited`.

nate-chandler committed:
[NFC] ReachableBlocks: Rename isReachable method.

It was previously `isVisited`.  And mark `visit` private, it needn't be
part of the interface for any current clients.

nate-chandler committed:
[NFC] ReachableBlocks: Add ::hasUnreachableBlocks.

The method walks over the blocks of the function testing each for
reachability and returns false on the first that is not.

nate-chandler committed:
[NFC] SGC: Extract range completion.

In preparation for completing lifetimes in multiple ranges (namely
post-orders of subgraphs of the function), extract a complete-in-range
method.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
Source compat failures match failures in baseline:
```
Failures:
  FAIL: SwiftLint-Legacy, 4.2, 60f98e, Swift Package
  FAIL: SwiftLint-Legacy, 4.0, 60f98e, Swift Package
```

eeckstein commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -104,39 +107,189 @@ struct SILGenCleanup : SILModuleTransform {
+
+/// Populate `roots` with the last blocks that are discovered via backwards
+/// walks along any non-repeating paths starting at the ends in `walks`.
+void collectReachableRoots(SILFunction *function, SmallVectorImpl<Walk> &walks,
```
This algorithm looks a bit complicated. Would this also work?

1. add blocks which are reachable from `unreachable` in BasicBlockSet _s_ (by going backwards in the CFG)
2. For each block: if the block is unreachable _and_ has no predecessors _and_ is in _s_: add it to the roots
3. add blocks which are reachable from roots in BasicBlockSet _t_ (by going forward in the CFG)
4. pick one unreachable block which is not in _t_ and add it to the root (to handle unreachable loops)
5. continue with 3. until all unreachable blocks are in _t_.

nate-chandler replied:
This sounds better, thanks!  Ill give it a shot.

nate-chandler replied:
Done except that a block in `s` but not `t` is not necessarily in an unreachable loop, only forwards-reachable from (at least) one, so it's necessary to walk backwards from such a block to find a representative block of the loop (see `unreachable_def_4`).  Also, the initial backward walk begins at all dead-end blocks; I was previously only starting at unreachable-terminated blocks which misses unreachable infinite loops.

nate-chandler commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -104,39 +107,189 @@ struct SILGenCleanup : SILModuleTransform {
+      }
+
+      // No predecessors?  Found a root.
+      if (block->pred_empty()) {
```
Yeah, unreachable loops (which are reachable walking backwards from `unreachable` instructions) are the reason for the complexity of the initial approach.  They're handled via `backedges`.

nate-chandler committed:
[NFC] BasicBlockWorklist: add ::empty.

Enable clients to check whether the worklist has anything in it without
popping a block to find out.

nate-chandler committed:
[SGC] Complete deadend-reachable lifetimes.

For a function to have complete lifetimes, the lifetime of every def in
the function which is backwards-reachable from a dead-end block must be
completed.

Previously, every def in every block which appears in the post-order of
the function's blocks is completed.  This was not enough: defs in
"unreachable blocks" (i.e. those which aren't forwards-reachable from
the function entry) must be completed too, and such blocks do not appear
in the function's post-order.

Here, defs in unreachable blocks are be completed too.  Such defs must
also be completed in relative post-order.  To do this, roots for the
non-entry post-orders must be found.

rdar://141197164

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

eeckstein commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -104,39 +108,219 @@ struct SILGenCleanup : SILModuleTransform {
+  //
+  // Every unpredecessored block reachable from some dead-end block is a root.
+  for (auto &block : *function) {
+    if (&block == function->getEntryBlock()) {
```
You don't need this condition because the entry block has no predecessors (which is checked below)

nate-chandler replied:
Not quite.  The code here is
```
    if (&block == function->getEntryBlock()) {
      continue;
    }
    if (!block.pred_empty())
      continue;
```
The entry block has no predecessors, so we don't continue below.

eeckstein commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -104,39 +108,219 @@ struct SILGenCleanup : SILModuleTransform {
+}
+
+/// Populate `roots` with the last blocks that are discovered via backwards
+/// walks along any non-repeating paths starting at the ends in `walks`.
```
`walks` -> `backward`

eeckstein commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -104,39 +108,219 @@ struct SILGenCleanup : SILModuleTransform {
+// Iterate over `iterator` until finding a block in `include` and not in
+// `exclude`.
+SILBasicBlock *findFirstBlock(SILFunction::iterator &iterator,
+                              BasicBlockSet const &include,
```
You could also directly pass `BasicBlockWorklist`s for `include` and `exclude`. Then you don't need to add `BasicBlockWorklist::getVisited()`

nate-chandler replied:
I prefer this approach.  I think this method should exist anyway.

eeckstein replied:
> I think this method should exist anyway.

I don't agree because it's an implementation detail of BasicBlockWorklist that `isVisited` is implemented as a set. Exposing this detail makes BasicBlockWorklist a leaking abstraction.

nate-chandler replied:
I don't think this constitutes any meaningful abstraction leakage (a worklist is going to use a set), and I think we it should be possible to get read-only access to the set that is produced via the worklist to pass off to other code that operates on sets.

nate-chandler replied:
I replaced uses of the sets with function_refs wrapping calls to isVisited.

eeckstein commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -104,39 +108,219 @@ struct SILGenCleanup : SILModuleTransform {
+  ReachableBlocks reachableBlocks(function);
+  reachableBlocks.compute();
+  StackList<SILBasicBlock *> roots(function);
+  if (!reachableBlocks.hasUnreachableBlocks()) {
```
It's probably not worth checking for the special case of no unreachable blocks in the function. `collectReachableRoots` will yield the same result with approximately the same effort.

eeckstein replied:
Well, you could do the work of loop #2 in loop #1, which reduces the overhead to a single loop, which is IMO negligible.

nate-chandler replied:
While it might be negligible, it's still worse and it's already written this way.  Why change it to something negligibly worse?

eeckstein replied:
The benefit would be to reduce the complexity of the source code (which is always my goal # 1): no need for `hasUnreachableBlocks`, less different branches to be taken, etc.

> it's still worse

I don't think so. We are iterating over blocks and instructions so many times that the overhead is for sure not measurable.

> it's already written this way

This is not a good argument 🙂

All that being said, this is nitpicking. If you prefer to keep it this way, I'm fine with it.

eeckstein reviewed (commented):
Nice!
Just some minor comments.

nate-chandler committed:
[NFC] ReachableBlocks: Use StackList.

Replace a SmallVector.

nate-chandler commented on include/swift/SILOptimizer/Utils/BasicBlockOptUtils.h:
```diff
@@ -38,10 +38,32 @@ class SILLoopInfo;
+  void compute();
+
+  /// Whether `block` is reachable from the entry block.
+  bool isReachable(SILBasicBlock *block) const {
```
Done.  Also, replaced the SmallVector in visit with a StackList.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
@swift-ci please test source compatibility

eeckstein reviewed (approved):
lgtm!

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test linux platform
